wsd/ProxyRequestHandler.cpp

```diff
@@ -0,0 +1,99 @@
+/* -*- Mode: C++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4; fill-column: 100 -*- */
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <config.h>
+
+#include <Poco/URI.h>
+
+#include <COOLWSD.hpp>
+#include "ProxyRequestHandler.hpp"
+#include <net/HttpRequest.hpp>
+#include <net/HttpHelper.hpp>
+
+std::map<unsigned long, std::shared_ptr<StreamSocket>> ProxyRequestHandler::LinkSession;
+std::map<std::string, std::shared_ptr<http::Response>> ProxyRequestHandler::CacheFileHash;
+std::chrono::system_clock::time_point ProxyRequestHandler::MaxAge;
+
+void ProxyRequestHandler::handleRequest(const std::string& relPath,
+                                        const std::shared_ptr<StreamSocket>& socket)
+{
+    Poco::URI uriProxy(ProxyServer);
+    const auto timeNow = std::chrono::system_clock::now();
+
+    if (MaxAge > std::chrono::system_clock::time_point() && timeNow > MaxAge)
+    {
+        CacheFileHash.clear();
+        MaxAge = std::chrono::system_clock::time_point();
+    }
+
+    const auto cacheEntry = CacheFileHash.find(relPath);
+    if (cacheEntry != CacheFileHash.end())
+    {
+        socket->sendAndShutdown(*cacheEntry->second.get());
+        return;
+    }
+
+    uriProxy.setPath(relPath);
+    auto sessionProxy = http::Session::create(uriProxy.getHost(),
+                                              http::Session::Protocol::HttpSsl,
+                                              uriProxy.getPort());
+    sessionProxy->setTimeout(std::chrono::seconds(10));
+    http::Request requestProxy(uriProxy.getPathAndQuery());
+    http::Session::FinishedCallback proxyCallback =
+        [](const std::shared_ptr<http::Session>& httpSession)
+            {
+                const auto callbackNow = std::chrono::system_clock::now();
+                auto sessionLink = LinkSession.find(reinterpret_cast<unsigned long>(httpSession.get()));
+                if (sessionLink == LinkSession.end())
+                {
+                    return;
+                }
+
+                try
+                {
+                    const std::shared_ptr<const http::Response> httpResponse = httpSession->response();
+                    if (httpResponse->statusLine().statusCode() == 200)
+                    {
+                        if (MaxAge == std::chrono::system_clock::time_point())
+                        {
+                            MaxAge = callbackNow + std::chrono::hours(10);
+                        }
+
+                        std::shared_ptr<http::Response> entry =
+                            std::make_shared<http::Response>(*httpResponse.get());
+                        CacheFileHash[httpSession->getUrl()] = entry;
```

Comment:
Since we aren't really using this `http::Response` as it was intended, and since we really don't want to share it at all, why not store the instance directly (no pointers). (Alternatively, we could use `std::unique_ptr`, again, since don't need the overhead of `shared_ptr`.)

I'd use `std::unordered_map`, because it has a much better locality than `std::map` (i.e. CPU cache and memory friendly), and should also be faster, on average. Once the cache is built, it's unlikely to change much. Then, we can simply do this: `CacheFileHash.emplace_back(httpSession->getUrl(), *httpResponse);` 
This will store a copy of the `Response` object directly in the map. Then when we have a cache-hit, we simply do: `socket->sendAndShutdown(cacheEntry->second);`

Reply:
The idea is moving only pointers rather than moving large data blob, I have some other ideas to simplify to moving pointers